EIPS/eip-7643.md

```diff
@@ -0,0 +1,1983 @@
+---
+eip: 7643
+title: History accumulator for pre-PoS data
+description: Commit to a single root for all block data before the merge.
+author: lightclient (@lightclient), kdeme (@kdeme)
+discussions-to: https://ethereum-magicians.org/t/eip-7643-history-accumulator-for-pre-pos-data/19014
+status: Draft
+type: Standards Track
+category: Core
+created: 2024-02-29
+---
+
+## Abstract
+
+Defines an SSZ object for accumulating all pre-PoS data and commit to the
+historical hashes accumulator's root
+`0xec8e040fd6c557b41ca8ddd38f7e9d58a9281918dc92bdb72342a38fb085e701`.
+
+## Motivation
+
+There are two main uses we consider for the historical hashes accumulator:
+
+* for users who wish to download the pre-PoS data for the execution chain and
+  verify it without executing each block, they may simply compute each block
+  hash, accumulate the epoch records, and then compare the local accumulator
+  root with the expected value.
+* additionally, the accumulator root allows for `O(log(n))` sized proofs to any
+  pre-PoS block, whereas today to achieve something similar, one must
+  recursively verify the header chain.
+
+## Specification
+
+### Historical Hashes Accumulator
+
+The historical hashes accumulator commits to the set of pre-merge headers and their
+associated total difficulty. The format for this data is defined as:
+
+```python
+EPOCH_SIZE = 8192 # blocks
+MAX_HISTORICAL_EPOCHS = 131072  # 2**17
+
+# An individual record for a historical header.
+HeaderRecord = Container[block_hash: bytes32, total_difficulty: uint256]
+
+# The records of the headers from within a single epoch
+EpochRecord = List[HeaderRecord, max_length=EPOCH_SIZE]
+
+HistoricalHashesAccumulator = Container[
+    historical_epochs: List[bytes32, max_length=MAX_HISTORICAL_EPOCHS],
+    current_epoch: EpochRecord,
+]
+```
+
+### Pre-PoS Root
+
+The hash tree root of `HistoricalHashesAccumulator` for data before block 15537393 is
+`0xec8e040fd6c557b41ca8ddd38f7e9d58a9281918dc92bdb72342a38fb085e701`.
```

Comment:
Can you mine a fancy number (by shifting the data structure around)? 😅

Reply:
unfortunately not a lot a of malleability in the SSZ object 😅